src/main/kotlin/org/rust/ide/docs/RsQualifiedNameProvider.kt

```diff
@@ -0,0 +1,20 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.docs
+
+import com.intellij.ide.actions.QualifiedNameProvider
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import org.rust.lang.core.psi.ext.RsQualifiedNamedElement
+import org.rust.lang.core.psi.ext.qualifiedName
+
+class RsQualifiedNameProvider : QualifiedNameProvider {
+    override fun getQualifiedName(element: PsiElement?): String? = (element as? RsQualifiedNamedElement)?.qualifiedName
+
+    override fun qualifiedNameToElement(fqn: String?, project: Project?): PsiElement? = null
```

Comment:
Just before I created #2544, I had introduced "psi links" feature in [EduTools](https://plugins.jetbrains.com/plugin/10081-edutools/) plugin (plugin for programming language courses in JetBrains IDEs)

In short, an educator can copy reference to psi element, add it to the description of course task, and when you click on it, it navigates you to the corresponding element in the editor.
![2020-05-22 13 38 55](https://user-images.githubusercontent.com/2539310/82659586-9b3cb900-9c31-11ea-9227-26923851da48.gif)

To get correct PSI element from text `QualifiedNameProvider#qualifiedNameToElement` is used.

Also, there is an [adaptation](https://stepik.org/course/59778) of [rustlings](https://github.com/rust-lang/rustlings) where this feature can be used. So I will think that #2544 is fully implemented when `qualifiedNameToElement` will be supported

But the current implementation of `getQualifiedName` doesn't allow us to do it because it produces the same names for different items sometimes. For example, 

```rust
struct Foo {
    x: i32 // fqn is crate_name::Foo::x
}

impl Foo {
    fn x(&self) -> i32 { self.x }  // fqn is crate_name::Foo::x
}
```

I had an idea to use `org.rust.lang.core.psi.ext.RsQualifiedName` to generate unique strings. 
It already can:
* be created from `RsQualifiedNamedElement` (`org.rust.lang.core.psi.ext.RsQualifiedName.Companion#from(org.rust.lang.core.psi.ext.RsQualifiedNamedElement)`
* provide unique name (see `org.rust.lang.core.psi.ext.RsQualifiedName#toUrlPath`). It's used to generate external documentation URLs so it uses `/` as separator but I think we can introduce similar method with `::` as separator to be more consistent with rust syntax
* be created from string `org.rust.lang.core.psi.ext.RsQualifiedName.Companion#from(java.lang.String)`
* almost find the corresponding psi element `org.rust.lang.core.psi.ext.RsQualifiedName#findPsiElement`. It should be improved because now it requires `context` to get workspace object

Reply:
So if I understand it correctly, the plan is to use `RsQualifiedName` to generate a unique reference path, which could be resolved back to the corresponding element from outside links, and change this unique path to something that can be resolved in Rust when you paste it in the editor?

Reply:
@Kobzol exactly!